### PR TITLE
Issue #1857 : Bcc recipient doesn't appear in drafts

### LINKED ISSUE #1857: Bcc recipient doesn't appear in drafts
### Expected behavior
When opening a draft, the user should be able to see the emails of the bcc field

### Actual behavior
Only the big letter with a coloured background appears instead of the email. Also the cursor position is wrong ( under the background of the letter)

### Steps to reproduce
1. Create a new message
2. Add a bcc address
3. Save as a draft
4. Open the draft and check the bcc field

### Environment
K-9 Mail version: 5.115

Android version:
4.4

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

### PR DESCRIPTION
**Pull request by utk-spartan**
Fixes #1857

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java [hunk 1]
```diff
@@ -202,7 +202,13 @@ public void setCcVisibility(boolean visible) {
     }
 
     public void setBccVisibility(boolean visible) {
+        boolean bccAlreadyVisible = isBccVisible();
         bccWrapper.setVisibility(visible ? View.VISIBLE : View.GONE);
+        if(visible && !bccAlreadyVisible){
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java]: This removes all recipients and re-adds them, fixing the symptom but not the cause. Do you know what causes the underlying problem?

Also, this is more logic than an MvpView class should have

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java [hunk 2]
```diff
@@ -241,7 +241,7 @@ public void addBccAddresses(Address... bccRecipients) {
             boolean alreadyVisible = recipientMvpView.isBccVisible();
             boolean singleBccRecipientFromAccount =
                     bccRecipients.length == 1 && bccRecipients[0].toString().equals(bccAddress);
-            recipientMvpView.setBccVisibility(alreadyVisible || singleBccRecipientFromAccount);
+            recipientMvpView.setBccVisibility(alreadyVisible || !singleBccRecipientFromAccount);
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: Flipped the logic here, was this actually wrong before?

Comment (utk-spartan) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: yup 
the bcc field is supposed to be hidden when the only address present is the default one; but instead it was hidden when it had addresses other than the default bcc address

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java (hunk 1 above)
Comment (utk-spartan) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientMvpView.java]: I agree, this a workaround not a fix.
It was caused by the tokenautocomplete external library which bugs out when the editbox containing  data is made visible.

I can put this whole removing and re-adding thing in the recipientPresenter class
or shall i work on a fix?

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java [hunk 3]
```diff
@@ -318,6 +325,7 @@ void onClickBccLabel() {
     void onClickRecipientExpander() {
         recipientMvpView.setCcVisibility(true);
         recipientMvpView.setBccVisibility(true);
+        if(defaultBccAddress != null) addBccAddresses(defaultBccAddress);
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: Code style - add a space after the `if` and always use a new line and `{}`'s.

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java [hunk 4]
```diff
@@ -234,14 +235,20 @@ private void addCcAddresses(Address... ccAddresses) {
...
-            addRecipientsFromAddresses(RecipientType.BCC, bccRecipients);
             String bccAddress = account.getAlwaysBcc();
 
             // If the auto-bcc is the only entry in the BCC list, don't show the Bcc fields.
             boolean alreadyVisible = recipientMvpView.isBccVisible();
             boolean singleBccRecipientFromAccount =
                     bccRecipients.length == 1 && bccRecipients[0].toString().equals(bccAddress);
-            recipientMvpView.setBccVisibility(alreadyVisible || singleBccRecipientFromAccount);
+            boolean showBcc = alreadyVisible || !singleBccRecipientFromAccount;
+            if(showBcc){
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: Space before and after ()'s

Comment (utk-spartan) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: done

File: k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java (hunk 3 above)
Comment (utk-spartan) [inline k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java]: done

**Comment by Valodim** [comment]
Does simply setting the view to visible on initialization if it's non-empty fix this issue? Keeping some recipient state in a different place seems like a dangerous workaround

**Comment by utk-spartan** [comment]
@Valodim Setting the view to visible on initialization if its non empty fixes the issue as i mentioned here https://github.com/k9mail/k-9/issues/1857#issuecomment-290074676 
but that means it will be shown even when it only contains the default bcc address, which is not we ought to do

**Comment by cketti** [comment]
I can't reproduce the issue with a current `master` build. I also don't really understand the fix.

**Pull request closed without merging**